cdap-watchdog/src/main/java/co/cask/cdap/logging/logbuffer/cleaner/LogBufferCleaner.java

```diff
@@ -0,0 +1,112 @@
+import co.cask.cdap.logging.meta.CheckpointManager;
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Runnable to clean up log buffer files for which logs are already persisted.
+ */
+public class LogBufferCleaner implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(LogBufferCleaner.class);
+  private final List<CheckpointManager<LogBufferFileOffset>> checkpointManagers;
+  private final File baseLogDir;
+
+  public LogBufferCleaner(CConfiguration cConf, List<CheckpointManager<LogBufferFileOffset>> checkpointManagers) {
+    this(checkpointManagers, cConf.get(Constants.LogBuffer.LOG_BUFFER_BASE_DIR));
+  }
+
+  @VisibleForTesting
+  LogBufferCleaner(List<CheckpointManager<LogBufferFileOffset>> checkpointManagers, String baseLogDir) {
+    this.checkpointManagers = checkpointManagers;
+    this.baseLogDir = new File(baseLogDir);
+  }
+
+  @Override
+  public void run() {
+    if (!baseLogDir.exists()) {
+      return;
+    }
+
+    try {
+      // From all the pipelines, get the largest file id for which logs have been persisted. So the cleaner can delete
+      // files with file id smaller than that file id.
+      long largestFileId = getLargestFileId(checkpointManagers);
+
+      // scan files under baseLogDir and delete file which has fileId smaller than largestFileId
+      scanAndDelete(baseLogDir, largestFileId);
+    } catch (Exception e) {
+      LOG.warn("Exception while cleaning up log buffer.", e);
+    }
+  }
+
+  /**
+   * Scans base log directory and deletes file which has fileId smaller than largestFileId.
+   */
+  private void scanAndDelete(File baseLogDir, long largestFileId) {
+    // if the offsets are not persisted yet, then nothing needs to be done.
+    if (largestFileId == -1) {
+      return;
+    }
+
+    File[] files = baseLogDir.listFiles();
+    if (files != null) {
+      int index = 0;
+      while (index < files.length) {
+        File file = files[index++];
+        if (getFileId(file.getName()) < largestFileId) {
+          // if the file is not deleted, in the next scan it should get deleted.
+          file.delete();
+        }
+      }
+    }
+  }
+
+  /**
+   * Get largest fileId for which logs have been persisted.
+   */
+  private long getLargestFileId(List<CheckpointManager<LogBufferFileOffset>> checkpointManagers) throws IOException {
+    // there will be atleast one log pipeline
+    LogBufferFileOffset minOffset = checkpointManagers.get(0).getCheckpoint(0).getOffset();
+
+    for (int i = 1; i < checkpointManagers.size(); i++) {
+      LogBufferFileOffset offset = checkpointManagers.get(i).getCheckpoint(0).getOffset();
+      // keep track of minimum offset of all the pipeline
+      minOffset = minOffset.compareTo(offset) > 0 ? offset : minOffset;
+    }
+
+    return minOffset.getFileId();
```

Comment:
What if this file is still being read from? Should this be `minOffset.getFileId() - 1`?

Reply:
Discussed offline with @poornachandra . Even though this case is handled by current code, the decision is to use file id - 1 so that its easy to understand. 

Fixed